WS2012R2/lisa/stateEngine.ps1

```diff
@@ -817,49 +824,157 @@ function DoSystemDown([System.Xml.XmlElement] $vm, [XML] $xmlData)
         }
         else
         {
-            $testData = GetTestData $vm.currentTest $xmlData
-            if ($testData -is [System.Xml.XmlElement])
+            UpdateState $vm $ApplyCheckpoint
+        }
+    }
+    else
+    {
+        #for NON-SUT VMs, put the VM state to run preStartScript or StartSystem
+        if ($vm.preStartConfig)
+        {
+            UpdateState $vm $RunSetupScript
+        }
+        else
+        {
+            UpdateState $vm $StartSystem
+        }
+    }
+}
+
+
+########################################################################
+#
+# DoApplyCheckpoint()
+#
+########################################################################
+function DoApplyCheckpoint([System.Xml.XmlElement] $vm, [XML] $xmlData)
+{
+    <#
+    .Synopsis
+        Apply checkpoint to let VM go into a know status if noCheckpoint=False.
+    .Description
+        Apply checkpoint if noCheckpoint=False. Then transition to RunSetupScript
+        if the currentTest defines a setup script. Otherwise, transition to StartSystem.
+        If noCheckpoint=True or not configured, no checkpoint is applied and do state
+        transition instead.
+    .Parameter vm
+        XML Element representing the VM under test.
+    .Parameter xmlData
+        XML document for the test.
+    .Example
+        DoSystemDown $testVM $xmlData
+    #>
+
+    if (-not $vm -or $vm -isnot [System.Xml.XmlElement])
+    {
+        LogMsg 0 "Error: DoApplyCheckpoint received an bad VM parameter"
+        return
+    }
+
+    LogMsg 9 "Info : Entering DoApplyCheckpoint( $($vm.vmName) )"
+
+    if (-not $xmlData -or $xmlData -isnot [XML])
+    {
+        LogMsg 0 "Error: DoApplyCheckpoint received a null or bad xmlData parameter - VM $($vm.vmName) disabled"
+        $vm.emailSummary += "    DoApplyCheckpoint received a null xmlData parameter - VM disabled<br />"
+        $vm.currentTest = "done"
+        UpdateState $vm $Disabled
+    }
+
+    $testData = GetTestData $vm.currentTest $xmlData
+    if ($testData -is [System.Xml.XmlElement])
+    {
+        # Do not need to recover from checkpoint
+        if ($testData.noCheckpoint -and $testData.noCheckpoint -eq "True")
```

Comment:
Also do not recover from checkpoint when $testData.noCheckpoint is null or empty. To avoid changing behavior of existing test cases.